File: game/scripts/vscripts/modifiers/global/intelect_magic_resist.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local __TS__Decorate = ____lualib.__TS__Decorate
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["8"] = 1,["9"] = 1,["10"] = 1,["11"] = 3,["12"] = 4,["13"] = 3,["14"] = 4,["15"] = 5,["16"] = 6,["17"] = 5,["18"] = 9,["19"] = 10,["20"] = 9,["21"] = 13,["22"] = 14,["23"] = 13,["24"] = 17,["25"] = 18,["26"] = 17,["27"] = 21,["28"] = 22,["29"] = 21,["30"] = 25,["31"] = 26,["32"] = 25,["33"] = 29,["34"] = 30,["35"] = 29,["36"] = 33,["37"] = 34,["38"] = 35,["39"] = 33,["40"] = 4,["41"] = 3,["42"] = 4,["44"] = 4});
local ____exports = {}
local ____dota_ts_adapter = require("utils.dota_ts_adapter")
local BaseModifier = ____dota_ts_adapter.BaseModifier
local registerModifier = ____dota_ts_adapter.registerModifier
____exports.modifier_intelect_magic_resist = __TS__Class()
local modifier_intelect_magic_resist = ____exports.modifier_intelect_magic_resist
modifier_intelect_magic_resist.name = "modifier_intelect_magic_resist"
__TS__ClassExtends(modifier_intelect_magic_resist, BaseModifier)
function modifier_intelect_magic_resist.prototype.IsHidden(self)
    return true
end
function modifier_intelect_magic_resist.prototype.IsPurgable(self)
    return false
end
function modifier_intelect_magic_resist.prototype.IsPurgeException(self)
    return false
end
function modifier_intelect_magic_resist.prototype.IsDebuff(self)
    return false
end
function modifier_intelect_magic_resist.prototype.RemoveOnDeath(self)
    return false
end
function modifier_intelect_magic_resist.prototype.AllowIllusionDuplicate(self)
    return true
end
function modifier_intelect_magic_resist.prototype.DeclareFunctions(self)
    return {MODIFIER_PROPERTY_MAGICAL_RESISTANCE_DIRECT_MODIFICATION}
end
function modifier_intelect_magic_resist.prototype.GetModifierMagicalResistanceDirectModification(self)
    local hero = self:GetParent()
    return hero:GetIntellect(true) * -0.05
end
modifier_intelect_magic_resist = __TS__Decorate(
    {registerModifier(nil)},
    modifier_intelect_magic_resist
)
____exports.modifier_intelect_magic_resist = modifier_intelect_magic_resist
return ____exports

File: src/vscripts/modifiers/global/intelect_magic_resist.ts
import { BaseModifier, registerModifier } from '../../utils/dota_ts_adapter';
import { calculateIntellectMagicResist } from './intellect-magic-resist-calculator';

@registerModifier('modifiers/global/intelect_magic_resist')
export class modifier_intelect_magic_resist extends BaseModifier {
  IsHidden(): boolean {
    return true;
  }

  IsPurgable(): boolean {
    return false;
  }

  IsPurgeException(): boolean {
    return false;
  }

  IsDebuff(): boolean {
    return false;
  }

  RemoveOnDeath(): boolean {
    return false;
  }

  AllowIllusionDuplicate(): boolean {
    return true;
  }

  DeclareFunctions(): ModifierFunction[] {
    return [ModifierFunction.MAGICAL_RESISTANCE_DIRECT_MODIFICATION];
  }

  GetModifierMagicalResistanceDirectModification(): number {
    const hero = this.GetParent() as CDOTA_BaseNPC_Hero;
    const intellect = hero.GetIntellect(true);

    // 原版：每点智力提供0.1%魔抗，基础25%，750智力时满（25% + 75% = 100%）
    // 解决方案：使用简单的分段线性函数，避免复杂计算
    // 低智力时扣减少，高智力时扣减多，确保魔抗不会轻易达到上限

    // 原版应该提供的魔抗（每点智力0.1%）
    const originalMagicResist = intellect * 0.001;

    // 计算实际应该获得的魔抗
    const actualMagicResist = calculateIntellectMagicResist(intellect);

    // 计算需要扣减的魔抗（原版魔抗 - 实际魔抗）
    // 返回负值，因为这是直接修改（扣减）
    return -(originalMagicResist - actualMagicResist);
  }
}
